File: ai-analytics-service/src/schemas/contract_registry.py
import json
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Dict, Literal

@dataclass
class ContractCapabilities:
    isSemanticContract: bool
    supportsDynamicQuestions: bool
    supportsBackgroundContext: bool
    supportsScoreDistribution: bool
    supportsPartialMaps: bool
    supportsAdaptationOutcome: bool
    usesStructuredDimensionSummary: bool
    usesNarrativeMetrics: bool
    hasOverallSummarySentenceLimit: bool
    stoneInterpretationSentenceLimit: Literal["2", "2-5"]


def load_contract_registry(data: Any) -> Dict[str, ContractCapabilities]:
    if not isinstance(data, dict) or not isinstance(data.get("versions"), dict):
        raise ValueError("Contract capabilities manifest must define versions")

    registry = {}
    for version, caps in data["versions"].items():
        if not isinstance(caps, dict):
            raise ValueError(f"Capabilities for {version} must be an object")
        try:
            registry[version] = ContractCapabilities(**caps)
        except TypeError as error:
            raise ValueError(f"Invalid capabilities for {version}: {error}") from error
    return registry


def _load_registry() -> Dict[str, ContractCapabilities]:
    # Walk up from src/schemas to project root, then down to contracts
    current_dir = Path(__file__).resolve().parent
    # current_dir is ai-analytics-service/src/schemas
    # project root is ai-analytics-service/..
    project_root = current_dir.parent.parent.parent
    capabilities_path = project_root / "contracts" / "capabilities.json"
    
    with capabilities_path.open("r", encoding="utf-8") as f:
        data = json.load(f)
        
    return load_contract_registry(data)


CONTRACT_REGISTRY = _load_registry()


def get_capabilities(version: str) -> ContractCapabilities:
    if version not in CONTRACT_REGISTRY:
        raise ValueError(f"Unknown contract version '{version}'")
    return CONTRACT_REGISTRY[version]

File: src/lib/contract-registry.ts
import capabilitiesManifest from '../../contracts/capabilities.json';

export interface ContractCapabilities {
  isSemanticContract: boolean;
  supportsDynamicQuestions: boolean;
  supportsBackgroundContext: boolean;
  supportsScoreDistribution: boolean;
  supportsPartialMaps: boolean;
  supportsAdaptationOutcome: boolean;
  usesStructuredDimensionSummary: boolean;
  usesNarrativeMetrics: boolean;
  hasOverallSummarySentenceLimit: boolean;
  stoneInterpretationSentenceLimit: '2' | '2-5';
}

const BOOLEAN_CAPABILITY_FIELDS = [
  'isSemanticContract',
  'supportsDynamicQuestions',
  'supportsBackgroundContext',
  'supportsScoreDistribution',
  'supportsPartialMaps',
  'supportsAdaptationOutcome',
  'usesStructuredDimensionSummary',
  'usesNarrativeMetrics',
  'hasOverallSummarySentenceLimit',
] as const;

export function loadContractRegistry(
  manifest: unknown,
): Record<string, ContractCapabilities> {
  if (!manifest || typeof manifest !== 'object' || Array.isArray(manifest)) {
    throw new Error('Contract capabilities manifest must be an object');
  }
  const versions = (manifest as { versions?: unknown }).versions;
  if (!versions || typeof versions !== 'object' || Array.isArray(versions)) {
    throw new Error('Contract capabilities manifest must define versions');
  }

  const registry: Record<string, ContractCapabilities> = {};
  for (const [version, rawCapabilities] of Object.entries(versions)) {
    if (
      !rawCapabilities ||
      typeof rawCapabilities !== 'object' ||
      Array.isArray(rawCapabilities)
    ) {
      throw new Error(`Capabilities for ${version} must be an object`);
    }
    const capabilities = rawCapabilities as Record<string, unknown>;
    for (const field of BOOLEAN_CAPABILITY_FIELDS) {
      if (typeof capabilities[field] !== 'boolean') {
        throw new Error(`Capabilities for ${version} require boolean ${field}`);
      }
    }
    if (
      capabilities.stoneInterpretationSentenceLimit !== '2' &&
      capabilities.stoneInterpretationSentenceLimit !== '2-5'
    ) {
      throw new Error(
        `Capabilities for ${version} have an invalid stone interpretation limit`,
      );
    }
    registry[version] = capabilities as unknown as ContractCapabilities;
  }
  return registry;
}

export const CONTRACT_REGISTRY = Object.freeze(
  loadContractRegistry(capabilitiesManifest),
);

export function getCapabilities(version: string): ContractCapabilities {
  const capabilities = CONTRACT_REGISTRY[version];
  if (!capabilities) {
    throw new Error(`Unknown contract version: ${version}`);
  }
  return capabilities;
}

export function isVersionSupported(version: string): boolean {
  return version in CONTRACT_REGISTRY;
}
